pkg/filesystem/resolve.go

```diff
@@ -0,0 +1,158 @@
+// * If path is a symlink, resolve it's target. If the target is not whitelisted add it to the
+// output set.
+// * Add all ancestors of each path to the output set.
+func ResolvePaths(paths []string, wl []util.WhitelistEntry) (pathsToAdd []string, err error) {
+	logrus.Info("Resolving paths")
+
+	fileSet := make(map[string]bool)
+
+	for _, f := range paths {
+		// If the given path is part of the whitelist ignore it
+		if util.IsInProvidedWhitelist(f, wl) {
+			logrus.Debugf("path %s is whitelisted, ignoring it", f)
+			continue
+		}
+
+		link, e := resolveSymlinkAncestor(f)
+		if e != nil {
+			err = e
+			return
+		}
+
+		if f != link {
+			logrus.Tracef("updated link %s to %s", f, link)
+		}
+
+		if !fileSet[link] {
+			pathsToAdd = append(pathsToAdd, link)
+		}
+		fileSet[link] = true
+
+		var evaled string
+
+		// If the path is a symlink we need to also consider the target of that
+		// link
+		evaled, err = filepath.EvalSymlinks(f)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				logrus.Errorf("couldn't eval %s with link %s", f, link)
+				return
+			}
+
+			logrus.Debugf("symlink path %s, target does not exist", f)
+		}
+
+		// If the given path is a symlink and the target is part of the whitelist
+		// ignore the target
+		if util.IsInProvidedWhitelist(evaled, wl) {
+			logrus.Debugf("path %s is whitelisted, ignoring it", evaled)
+			continue
+		}
+
+		if !fileSet[evaled] {
+			pathsToAdd = append(pathsToAdd, evaled)
+		}
+		fileSet[evaled] = true
+	}
+
+	// Also add parent directories to keep the permission of them correctly.
+	pathsToAdd = filesWithParentDirs(pathsToAdd)
+
+	return
+}
+
+// filesWithParentDirs returns every ancestor path for each provided file path.
+// I.E. /foo/bar/baz/boom.txt => [/, /foo, /foo/bar, /foo/bar/baz, /foo/bar/baz/boom.txt]
+func filesWithParentDirs(files []string) []string {
+	filesSet := map[string]bool{}
+
+	for _, file := range files {
+		file = filepath.Clean(file)
+		filesSet[file] = true
+
+		for _, dir := range util.ParentDirectories(file) {
+			dir = filepath.Clean(dir)
+			filesSet[dir] = true
+		}
+	}
+
+	newFiles := []string{}
+	for file := range filesSet {
+		newFiles = append(newFiles, file)
+	}
+
+	return newFiles
+}
+
+// resolveSymlinkAncestor returns the ancestor link of the provided symlink path or returns the
+// the path if it is not a link. The ancestor link is the filenode whose type is a Symlink.
+// E.G /baz/boom/bar.txt links to /usr/bin/bar.txt but /baz/boom/bar.txt itself is not a link.
+// Instead /bar/boom is actually a link to /usr/bin. In this case resolveSymlinkAncestor would
+// return /bar/boom.
+func resolveSymlinkAncestor(path string) (string, error) {
+	if !filepath.IsAbs(path) {
+		return "", errors.New("dest path must be abs")
+	}
+	last := ""
+	newPath := path
+loop:
+	for newPath != "/" {
+		fi, err := os.Lstat(newPath)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to lstat")
```

Comment:
does the err contain the file name? if not, please add the file name

Reply:
yes, lstat errors always contain the file name